Rework how existing LXD PIDs are validated/discovered

simondeziel opened the pull request:
Change how LXD's `daemon.start` detects existing LXD processes left behind.

1. Optimize how `pgrep` is used to avoid needing to invoke it twice, once for regular daemons and another for debug ones.
2. Tweak `pgrep` invocation to only report PIDs that have an EUID of `0`. This ensures we don't accidentally get nested LXD PIDs because those should be in a different userns (for `unpriv`)
3. Once we have a candidate PID, make sure it's from a snap (`SNAP_NAME=lxd` line in `/proc/$pid/environ`)
4. Now that we have a probable LXD PID, make sure it is really a leftover process and not a nested LXD from a `privileged` container.

This last step involves writing short random data through the PID's mount namespace (`/proc/${pid}/root/var/snap/lxd/common/lxd/.validate`) and comparing it with our own view of the mount namespace by accessing the same file but directly (`/var/snap/lxd/common/lxd/.validate`).

simondeziel committed:
daemon.start: improve grep inspecting LXD environment

Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

simondeziel committed:
daemon.start: only call pgrep once

Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

simondeziel committed:
daemon.start: explain how existing PIDs are validated

Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

simondeziel committed:
daemon.start: enhance how existing LXD PIDs are validated

By using some cheap random data to write and compare, we reduce the risk of a
nested LXD running the same code at the same time to also do this same
validation.

It is no longer needed to remove the .validate file first as we rely on its
content more than its presence.

Suggested-by: Alexander Mikhalitsyn <aleksandr.mikhalitsyn@canonical.com>
Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

simondeziel committed:
daemon.start: limit pgrep to match on root-owned LXD PIDs

This is a safeguard to ignore (unprivileged) nested LXD. Here is a comparison of the old way:

```
  server_pid: 1293
5923 /var/snap/lxd/common/lxd.debug --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd
8424 lxd --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd
```

In the above, the PID `8424` is the LXD inside the `c1` instance.

Now with the new way:

```
5923 /var/snap/lxd/common/lxd.debug --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd
```

Note: privileged nested LXD would also be running as EUID 0 but would be in a
different mount namespace which is what the `.validate` file comparison guards
against.

Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

tomponline commented on snapcraft/commands/daemon.start:
```diff
@@ -620,7 +620,8 @@ fi
 for pid in $(pgrep -f "lxd --logfile" ; pgrep -f "lxd.debug --logfile"); do
-    grep -qF "SNAP_NAME=lxd" "/proc/${pid}/environ" || continue
+    # Cheap confirmation that we likely have a LXD PID
+    grep -qxFm1 "SNAP_NAME=lxd" "/proc/${pid}/environ" || continue
```
Are there long form arguments we can use here? Either that or an extension of the commit message/comment to explain how this is an improvement/cheaper?

simondeziel replied:
While they all have long form versions, I think switching them all would make it harder to understand to some as most of those options I've only ever seen used in their short form.

I'm assuming the `-m1` bit is the one needing clarification here which I'm happy to expand as indeed it's not commonly used and the long form is quite clear (`--max-count=1`):
>        -m NUM, --max-count=NUM
>              Stop  reading  a  file  after  NUM matching lines.

tomponline replied:
actually it was most of them :)

tomponline replied:
but the use of --max-count makes sense to me now

simondeziel replied:
I've now expanded `-x` and `-F` as well.

tomponline commented on snapcraft/commands/daemon.start:
```diff
@@ -619,7 +619,7 @@ fi
 # LXD
 ## Check for existing LXDs
-for pid in $(pgrep -f "lxd(\.debug)? --logfile"); do
+for pid in $(pgrep --euid 0 --full "lxd(\.debug)? --logfile"); do
```
in the example in your commit, why is server_pid 1293, but this command doesn't match that process?

tomponline replied:
Also why not --uid?

simondeziel replied:
The `server_pid 1293` is relative to the container's PID NS which is why it doesn't match the host's view.

simondeziel replied:
As to why not `--uid` instead of `--euid` it shouldn't really matter here as they should both be `0`. The distinction becomes relevant when using setuid binaries.

simondeziel replied:
Also, in other scripts we use `id -u` to ensure the EUID is `0` so this should be more consistent.

tomponline commented on snapcraft/commands/daemon.start:
```diff
@@ -625,12 +625,11 @@ for pid in $(pgrep -f "lxd(\.debug)? --logfile"); do
-    rm -f "/var/snap/lxd/common/lxd/.validate"
-    touch "/proc/${pid}/root/var/snap/lxd/common/lxd/.validate" 2>/dev/null || true
-    if [ -e "/var/snap/lxd/common/lxd/.validate" ]; then
+    echo "$$" > "/proc/${pid}/root/var/snap/lxd/common/lxd/.validate" 2>/dev/null || true
```
I'm not really following why this change is needed (leaving aside the microoptimisation for now too).

What issue is this fixing? Have we seen it occur? 

I'm confused why a nested LXD would be able to write to the /var/snap/lxd/common/lxd directory of its parent?

simondeziel replied:
It's not fixing any known/observed issue more a theoretical/potential one.

This came about after I asked Aleks if it was OK to replace the touch+rm by just checking for `"/proc/${pid}/root/var/snap/lxd/common/lxd/server.key`'s presence.

Aleks quickly pointed out that would defeat the purpose of that check which is to make sure we are dealing with a LXD PID that belongs to our PID and mount NS. That's because any nested LXD PID we'd get from that `pgrep` would also lead us to a `server.key` but in a different mount NS. Hence the writing to it was needed.

I then observed that if the nested LXD was also going through that same validation at the same time, it could also race and create the `.validate` file to causing problems.

To which Aleks replied: "write and compare random data to eliminate the race".

simondeziel replied:
The micro-optimization was dropped.

tomponline commented:
@simondeziel it'd be interesting to know the context of the precursor discussions you may have had with @mihalicyn about this (he's referenced in the commits) as im not really aware of the overall reason for this PR.

Adding a PR description would likely help to add context here.

simondeziel committed:
daemon.start: always remove the .validate file

Doing the removal only on confirmed to be conflicting LXD meant that
potentially multiple files could be left behind, one per invalid PID.

Signed-off-by: Simon Deziel <simon.deziel@canonical.com>

simondeziel commented:
@tomponline I've updated the comment based on our discussion to better explain that the `.validate` dance is to rule out *priv* nested LXD by making sure the PID in question doesn't use the same mount namespace as "ours".

I've also reintroduced the removal of that `.validate` file but differently to avoid "leaking" some of them on inside directories of "failed" candidate LXD.
